spock-core/src/main/java/org/spockframework/mock/ClosureParameterTypeFromVariableType.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.spockframework.mock;
+
+import groovy.transform.stc.SingleSignatureClosureHint;
+import org.codehaus.groovy.ast.ASTNode;
+import org.codehaus.groovy.ast.ClassNode;
+import org.codehaus.groovy.ast.CodeVisitorSupport;
+import org.codehaus.groovy.ast.MethodNode;
+import org.codehaus.groovy.ast.expr.DeclarationExpression;
+import org.codehaus.groovy.ast.expr.Expression;
+import org.codehaus.groovy.ast.expr.MethodCallExpression;
+import org.codehaus.groovy.control.CompilationUnit;
+import org.codehaus.groovy.control.SourceUnit;
+import org.spockframework.compiler.AstUtil;
+import spock.mock.MockingApi;
+
+import java.util.*;
+
+public class ClosureParameterTypeFromVariableType extends SingleSignatureClosureHint {
```

Comment:
Is there some way to test this?

Reply:
To be honest I don't remember anymore how I tested it, but I'm pretty sure I did.
I wouldn't be able to write that code blindly. :-D
In IntelliJ it is not testable due to the bug https://youtrack.jetbrains.com/issue/IDEA-331844 which when resolved would benefit from this.

One way to test it is:
- open MockingApi
- copy the `Stub` method with only `Closure` as argument two times as `Stub2` and `Stub3` (to prevent substitution by `StubImpl`)
- Remove the `ClosureParams` annotation from  `Stub3` parameter
- Add `Stub2` and `Stub3` to the `MOCK_METHODS` in the closure hint implementation
- Start the `groovyConsole` with those changes

Now try to compile
```groovy
@groovy.transform.TypeChecked
class Foo extends spock.lang.Specification {
  def "w/o data provider"() {
    given:
    FeatureInfo featureInfo2 = Stub2 {
      it.getDataProcessorMethod()
    }
    FeatureInfo featureInfo3 = Stub3 {
      it.getDataProcessorMethod()
    }

    expect:
    true
  }
}

interface FeatureInfo {
  String getDataProcessorMethod()
}
```
and you will see that it complains for `Stub3` but not for `Stub2` as it knows the type of `it` is `FeatureInfo` in that case.